File: product_of_array_except_self.rb
# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
  left_prods = [1]
  right_prods = [1]
  lp = 1
  rp = 1
  count = nums.size
  
  (1...count).each do |i|
    lp *= nums[i-1]
    left_prods << lp
    
    rp *= nums[count - i]
    right_prods << rp
  end
  
  (0...count).map do |i|
    left_prods[i] * right_prods[count - i -1]
  end
end

File: product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums):
        n = len(nums)
        prods = []
        left_prods = [1]
        right_prods = [1]
        last_left_prod = 1
        last_right_prod = 1
        
        for i in range(1, n):
            last_left_prod *= nums[i - 1]
            left_prods.append(last_left_prod)

            last_right_prod *= nums[n - i]
            right_prods.append(last_right_prod)

        right_prods.reverse()

        for i in range(n):
            prods.append(left_prods[i] * right_prods[i])

        return prods


print(Solution().productExceptSelf([1, 2, 3, 4]))
